File: backend/src/main/java/cz/forgottenempire/servermanager/steamauth/SteamAuthController.java
package cz.forgottenempire.servermanager.steamauth;

import cz.forgottenempire.servermanager.api.SteamAuthApi;
import cz.forgottenempire.servermanager.api.model.AuthStatus;
import cz.forgottenempire.servermanager.api.model.AuthType;
import cz.forgottenempire.servermanager.api.model.AuthVerificationResultDto;
import cz.forgottenempire.servermanager.api.model.SteamAuthDto;
import cz.forgottenempire.servermanager.api.model.SteamAuthStatusDto;
import cz.forgottenempire.servermanager.security.permission.PermissionCode;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.RestController;

@RestController
@PreAuthorize("hasAuthority('" + PermissionCode.STEAM_AUTH_ADMIN + "')")
public class SteamAuthController implements SteamAuthApi {

    private final SteamAuthService authService;
    private final SteamAuthVerifier authVerifier;

    @Autowired
    public SteamAuthController(SteamAuthService authService, SteamAuthVerifier authVerifier) {
        this.authService = authService;
        this.authVerifier = authVerifier;
    }

    @Override
    public ResponseEntity<Void> setSteamAuth(SteamAuthDto steamAuthDto) {
        authService.setAuthAccount(steamAuthDto);
        return ResponseEntity.ok().build();
    }

    @Override
    public ResponseEntity<SteamAuthDto> getSteamAuth() {
        SteamAuth steamAuth = authService.getAuthAccount();
        SteamAuthDto dto = new SteamAuthDto()
                .username(steamAuth.getUsername())
                .steamGuardToken(steamAuth.getSteamGuardToken());
        return ResponseEntity.ok(dto);
    }

    @Override
    public ResponseEntity<Void> clearSteamAuth() {
        authService.clearAuthAccount();
        return ResponseEntity.noContent().build();
    }

    @Override
    public ResponseEntity<SteamAuthStatusDto> getSteamAuthStatus() {
        boolean isConfigured = authService.isAuthConfigured();
        return ResponseEntity.ok(new SteamAuthStatusDto().isConfigured(isConfigured));
    }

    @Override
    public ResponseEntity<AuthVerificationResultDto> verifySteamAuth(SteamAuthDto steamAuthDto) {
        try {
            AuthVerificationResult result = authVerifier.verifyCredentials(steamAuthDto);
            AuthVerificationResultDto dto = new AuthVerificationResultDto()
                    .status(AuthStatus.fromValue(result.getStatus().name()))
                    .message(result.getMessage())
                    .authType(AuthType.fromValue(result.getAuthType().name()));
            return ResponseEntity.ok(dto);
        } catch (Exception e) {
            AuthVerificationResultDto dto = new AuthVerificationResultDto()
                    .status(AuthStatus.ERROR)
                    .message("Failed to verify credentials: " + e.getMessage())
                    .authType(AuthType.UNKNOWN);
            return ResponseEntity.internalServerError().body(dto);
        }
    }
}

File: backend/src/main/kotlin/cz/forgottenempire/servermanager/steamauth/SteamAuthController.kt
package cz.forgottenempire.servermanager.steamauth

import cz.forgottenempire.servermanager.api.SteamAuthApi
import cz.forgottenempire.servermanager.api.model.SessionStatus
import cz.forgottenempire.servermanager.api.model.SteamAuthDto
import cz.forgottenempire.servermanager.api.model.SteamAuthStatusDto
import cz.forgottenempire.servermanager.api.model.SteamLoginRequestDto
import cz.forgottenempire.servermanager.api.model.SteamLoginResultDto
import org.springframework.http.ResponseEntity
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.RestController

private const val STEAM_AUTH_PERMISSION = "hasAuthority('STEAM_AUTH_ADMIN')"

@RestController
@PreAuthorize(STEAM_AUTH_PERMISSION)
class SteamAuthController(
    private val authService: SteamAuthService,
    private val loginService: SteamLoginService,
    private val sessionStatusHolder: SteamSessionStatusHolder
) : SteamAuthApi {

    override fun steamLogin(steamLoginRequestDto: SteamLoginRequestDto): ResponseEntity<SteamLoginResultDto> {
        return try {
            val result = loginService.login(
                steamLoginRequestDto.username,
                steamLoginRequestDto.password,
                steamLoginRequestDto.steamGuardCode
            )
            val dto = SteamLoginResultDto()
                .result(result.result)
                .authType(result.authType)
                .message(result.message)
            ResponseEntity.ok(dto)
        } catch (e: Exception) {
            val dto = SteamLoginResultDto()
                .result(cz.forgottenempire.servermanager.api.model.SteamLoginResult.ERROR)
                .message("Login failed: ${e.message}")
            ResponseEntity.internalServerError().body(dto)
        }
    }

    override fun getSteamAuth(): ResponseEntity<SteamAuthDto> {
        val steamAuth = authService.getAuthAccount()
        val dto = SteamAuthDto().username(steamAuth.username)
        return ResponseEntity.ok(dto)
    }

    override fun clearSteamAuth(): ResponseEntity<Void> {
        authService.clearAuthAccount()
        return ResponseEntity.noContent().build()
    }

    override fun getSteamAuthStatus(): ResponseEntity<SteamAuthStatusDto> {
        val isConfigured = authService.isAuthConfigured()
        val sessionStatus = if (!isConfigured) SessionStatus.NOT_CONFIGURED else sessionStatusHolder.status
        val dto = SteamAuthStatusDto()
            .isConfigured(isConfigured)
            .sessionStatus(sessionStatus)
            .lastCheckedAt(sessionStatusHolder.lastCheckedAt)
        return ResponseEntity.ok(dto)
    }
}
